ddtrace/contrib/celery/utils.py

```diff
@@ -1,57 +1,59 @@
+from typing import Any
+from typing import Dict
 from weakref import WeakValueDictionary
 
+from ddtrace.span import Span
+
 from .constants import CTX_KEY
 
 
-def tags_from_context(context):
+TAG_KEYS = frozenset(
+    [
+        ("compression", "celery.compression"),
+        ("correlation_id", "celery.correlation_id"),
+        ("countdown", "celery.countdown"),
+        ("delivery_info", "celery.delivery_info"),
+        ("eta", "celery.eta"),
+        ("exchange", "celery.exchange"),
+        ("expires", "celery.expires"),
+        ("hostname", "celery.hostname"),
+        ("id", "celery.id"),
+        ("priority", "celery.priority"),
+        ("queue", "celery.queue"),
+        ("reply_to", "celery.reply_to"),
+        ("retries", "celery.retries"),
+        ("routing_key", "celery.routing_key"),
+        ("serializer", "celery.serializer"),
+        ("timelimit", "celery.timelimit"),
+        # Celery 4.0 uses `origin` instead of `hostname`; this change preserves
+        # the same name for the tag despite Celery version
+        ("origin", "celery.hostname"),
+        ("state", "celery.state"),
+    ]
+)
+
+
+def set_tags_from_context(span, context):
+    # type: (Span, Dict[str, Any]) -> None
     """Helper to extract meta values from a Celery Context"""
-    tag_keys = (
-        "compression",
-        "correlation_id",
-        "countdown",
-        "delivery_info",
-        "eta",
-        "exchange",
-        "expires",
-        "hostname",
-        "id",
-        "priority",
-        "queue",
-        "reply_to",
-        "retries",
-        "routing_key",
-        "serializer",
-        "timelimit",
-        "origin",
-        "state",
-    )
-
-    tags = {}
-    for key in tag_keys:
+
+    for key, tag_name in TAG_KEYS:
         value = context.get(key)
 
         # Skip this key if it is not set
         if value is None or value == "":
             continue
 
-        # Skip `timelimit` if it is not set (it's default/unset value is a
+        # Skip `timelimit` if it is not set (its default/unset value is a
         # tuple or a list of `None` values
-        if key == "timelimit" and value in [(None, None), [None, None]]:
+        if key == "timelimit" and all(_ is None for _ in value):
             continue
 
-        # Skip `retries` if it's value is `0`
+        # Skip `retries` if its value is `0`
         if key == "retries" and value == 0:
             continue
 
-        # Celery 4.0 uses `origin` instead of `hostname`; this change preserves
-        # the same name for the tag despite Celery version
-        if key == "origin":
-            key = "hostname"
-
-        # prefix the tag as 'celery'
-        tag_name = "celery.{}".format(key)
-        tags[tag_name] = value
-    return tags
+        span.set_tag(tag_name, value)
```

Comment:
is this a place where `span._set_tag` could be used to optimize a bit as well? at least for values that are known to be stringify-able ?

Reply:
In principle yes; we could split the `TAG_KEYS` set into stringifiable, metrics and generic I suppose. WDYT?

Reply:
I know that `_set_tag` is a performance optimization when we know we have at least an `Optional[str]`, wondering if that would make a difference here or not?

(I'll be honest, I am not 100% sure where the optimization is here, we are doing one loop to extract + set rather than extract all and then set all?)

Reply:
> (I'll be honest, I am not 100% sure where the optimization is here, we are doing one loop to extract + set rather than extract all and then set all?)

Extraction came with the cost of allocating memory for a `dict`. The proposed change avoids this by setting the tags as they are extracted. There is also the bonus of pre-computing tag names instead of concatenating at runtime. But you are right, here we are just improving on some multiplicative coefficients, not complexities.